### DISCUSSION TITLE
Presets system - understanding the code

### DISCUSSION THREAD

**Discussion by philstopford** [CineEncoder/cine-encoder#76] [Q&A]
I'm breaking this out from other threads to make it easier to discuss. In working through the code to implement hard-burn subtitles options, allow Qt6 to be used, and adding multithread options, I still cannot understand the preset system and how the UI works with it.

- I don't understand where/how the initial presets are defined.
- I don't understand how the preset UI is set based on the preset definition. It seems like some kind of tree is navigated, but this approach is odd because I don't see that approach elsewhere in the tool.
- I'm not clear why this approach was chosen vs something like XML, which would allow versioning, backward/forward compatibility, etc.

I ran into a number of pain points when trying to add hard-burn settings to presets, which is why the above arise. The code seems rather convoluted to my eye and I wondered whether there might be a better / more accessible way to implement this system.

**Comment by philstopford** [comment]
@CineEncoder : just nudging this. I will try again shortly to make sense of this, but was hoping for a hint or two :)

**Comment by philstopford** [comment]
I've spent some time digging into this. Greater understanding came from writing out XML for the current parameters and preset table. My plan would be to work to replace the current Qt-provided ini 'blob' with XML. That might make it easier to work with and provide compatibility of settings between versions. Still something of a work-in-progress and I'll try and send over a PR once I've got it worked out.

**Comment by philstopford** [comment]
I have a PR here (https://github.com/CineEncoder/cine-encoder/pull/77) to replace the ini-hosted preset system with one based on XML. This also addresses the issue of presets being reset when updating versions (https://github.com/CineEncoder/cine-encoder/issues/25) because the XML tags allow for managing new properties quite nicely.
The PR does not touch settings.ini, but I would like to move that to XML as well, for similar reasons - having settings reset when updating is kind of annoying.

**Comment by CineEncoder** [comment]
Hello! I apologize for the delay. Default presets are defined in this file:
https://github.com/CineEncoder/cine-encoder/blob/master/app/resources/data/default_presets.ini
and is included here: https://github.com/CineEncoder/cine-encoder/blob/master/app/mainwindow.cpp#L2334
and here: https://github.com/CineEncoder/cine-encoder/blob/master/app/mainwindow.cpp#L2479
At that time, this recording method seemed more convenient to me due to the fact that the contents of the containers are directly written to a file using QDataStream. 
I merged some pull requests, some contain conflicts.
